src/jobs/shellcheck.yml

```diff
@@ -0,0 +1,95 @@
+description: |
+  Automatically shellcheck the run statements within your jobs and commands.
+
+executor: cli/default
+
+parameters:
+  source-dir:
+    description: Path containing the orb source files.
+    type: string
+    default: "src"
+  exclude:
+    description: >
+      Enter a comma-separated list, with no spaces, of errors to ignore. Example: "SC1009,SC1073"
+    type: string
+    default: ""
+
+steps:
+  - checkout
+  - run:
+      name: Install shellcheck
+      command: |
+        if [[ $EUID == 0 ]]; then export SUDO=""; else export SUDO="sudo"; fi
+        $SUDO apt-get install shellcheck
+  - run:
+      name: Install YQ
+      command: |
+        if [[ $EUID == 0 ]]; then export SUDO=""; else export SUDO="sudo"; fi
+        wget https://github.com/mikefarah/yq/releases/download/3.2.1/yq_linux_386
+        $SUDO mv yq_linux_386 /usr/local/bin/yq
+        $SUDO chmod +x /usr/local/bin/yq
+  - run:
+      name: ShellCheck
+      command: |
+        set +e
+        # Set colors for terminal output
+        term_red=$'\e[1;31m'
+        term_green=$'\e[1;32m'
+        term_end=$'\e[0m'
+
+        ORB_SOURCE_FILES="<< parameters.source-dir >>/**/*.yml"
+        SC_ERROR_COUNT=0
+
+        # EXCLUSIONS
+        # SC2148 - Missing shebang, this is to be expected.
+        # SC1044 - "\<<" parameter notation before the text replacement will appear like a here document with no EOF. Ignore this error for now.
+        # SC1072 - Not entirely sure on this one: https://github.com/koalaman/shellcheck/wiki/SC1072. Temporarily adding.
+        # SC1090 - https://github.com/koalaman/shellcheck/wiki/SC1090
+
+        EXCLUDED_SHELL_ERRORS=SC2148,SC1044,SC1072,SC1090<<#parameters.exclude>>,<</parameters.exclude>><<parameters.exclude>>
+
+        ###
+        # For each file
+        ###
+        for file in $ORB_SOURCE_FILES
+        do
+          # Get the number of steps
+          echo -----
+          echo "Scanning $file"
+          ORB_STEP_COUNT=$(yq r $file --length steps)
+          i=0
+          while [[ $i -lt $ORB_STEP_COUNT ]]
+            ###
+            # For each step in that file
+            ###
+          do
+            if [[ $(yq r $file steps[$i].run.command) ]]; then
+              ###
+              # If that step is a "run" step with a "command" key
```

Comment:
Problem here. This detects steps with the full/normal structure to run Bash commands. Alternatively, CircleCI accepts and many people run steps with the shorthand version:

```
steps:
  - run:
      command: echo "This is the standard way to run a command."
  - run: echo "This is the shorthand way to run a command."
```

Reply:
My thinking was if it was on a single line, you may not need to shell check it. 

However, we could still do this. another concern I was already having was if someone changes the default shell.

So I am thinking maybe change the logic so this is the new criteria of what is included.

- Any `run` step with no child keys (addresses your comment)
- Any `run` step with a `command` key that _does not_ also have a `shell` key.

Reply:
> My thinking was if it was on a single line, you may not need to shell check it.

Well either example can be multiple line:

```
steps:
  - run:
      command: |
        echo "This is the standard way to run a command."
        ./do-something
  - run: |
      echo "This is the shorthand way to run a command."
      ./do-something
```

> if someone changes the default shell

That's a really good point. That'll be hard to check. Even if they don't set "shell", the container could run Bash but it could still be something else.

Reply:
Ahh I forgot you could multiline the run command directly, great point.

If the job level shell is changed that could potentially be an issue. Should there be a file exclusion list?

Reply:
Improved the logic here overall. Now skips all non run-statements and run statements with a shell, will error if the command is missing (same effect as linting), and works on single run statements.

Failing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/104/workflows/ce893e63-1128-4c19-9658-63759014d628/jobs/319

Passing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/107/workflows/97f8f39c-74ec-40d4-bc47-7b6e10d48568/jobs/322

Reply:
I'm not sure. I haven't seen many projects change the shell.